[5.1] Guided Tours - Add checkbox/radio/select target element support, enhanced required field handling

GeraintEdwards opened the pull request:
Pull Request for Issue # .

### Summary of Changes

- Add checkbox/radio/select target element support for interactive field types
- Add support for textarea target element alongside input text type targets
- Allow fields that are not 'required' in the DOM to be required in the field e.g. for tours relating to searching where the input field is required for the tour but not by Joomla or the active page component
- Allow the tour to insist on a specific required value (translatable) so that the context of the tour can be maintained as it progresses e.g. searching for a specific text value.

Please note that a new database field has been added in this PR 'params' in the guided tours steps table.

### Testing Instructions

Test the provided tour.
Check that old tours still function properly.

### Actual result BEFORE applying this Pull Request

The tour will not respond to the checkbox, select list and textarea field changes correctly

### Expected result AFTER applying this Pull Request

See the tour as it plays out in these screenshots

![Screenshot from 2023-06-21 10-25-08](https://github.com/joomla/joomla-cms/assets/839810/92af8af1-b1e3-46a7-b8b1-fbe673f8ccbe)

Title: Global config demo tour
URL: ```/administrator/index.php?option=com_config```
Description: 
```
<p>This tour will demonstrate:</p>
<ul>
<li>checkboxes</li>
<li>radio boxes</li>
<li>required text fields searching for specific value</li>
</ul>
```

![Screenshot from 2023-06-21 10-25-13](https://github.com/joomla/joomla-cms/assets/839810/763dd30e-b7e1-41ca-8e6d-27177026dd2e)

Step 1

![Capture d'écran 2024-02-02 174546](https://github.com/joomla/joomla-cms/assets/5964177/2e08765d-2fc7-4fc7-aea5-5c319c8b94b3)

Title: Change editor to CodeMirror
Description: You should change the default editor and select 'CodeMirror'
Target: ```#jform_editor```
Type: interactive
Interactive Type: select list
Options: value required and value to enter: codemirror.

Note: all steps should have the position 'bottom'.

![Screenshot from 2023-06-21 10-25-18](https://github.com/joomla/joomla-cms/assets/839810/39fa0b94-cf24-43f9-a685-72521e8eda0c)

![Screenshot from 2023-06-21 10-25-22](https://github.com/joomla/joomla-cms/assets/839810/b8c97940-7d21-43ad-aaf1-3279671fdaad)

Step 2

Title: Put the site offline
Description: Now put the site offline
Target: ```#jform_offline1```
Type: interactive
Interactive Type: checkbox/radio
Options: value required.

![Screenshot from 2023-06-21 10-25-27](https://github.com/joomla/joomla-cms/assets/839810/683aac4b-69b6-452b-bcdb-d8b5b8aa9b85)

![Screenshot from 2023-06-21 10-25-33](https://github.com/joomla/joomla-cms/assets/839810/ec8c9486-9656-45dc-98a4-38a2bfc68403)

Step 3

Title: Change custom message
Description: Please set the custom message to: We will be back soon.
Target: ```#jform_offline_message```
Type: interactive
Interactive Type: text field
Options: value required and value to enter: ```We will be back soon.```

![Screenshot from 2023-06-21 10-25-47](https://github.com/joomla/joomla-cms/assets/839810/48e24410-1f58-472c-9b6a-a55be778822f)

![Screenshot from 2023-06-21 10-25-53](https://github.com/joomla/joomla-cms/assets/839810/5eadb707-e324-4ac6-b291-69f1da432fd8)

Step 4

Title: Check the frontend of the site
Description: The tour is now finished - save the configuration and check the site in the frontend



### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

Geraint2 committed:
New Radio/Checkbox option for interactive target type
Add required option for text and radio/checkbox target types to allow us to force the user to interact with the target field even if its not required in the DOM

Geraint2 committed:
Put advanced options on second tab in step editing

hans2103 committed:
handle required value for tests on text type input fields

Geraint2 committed:
handle required value for tests on text type input fields - store data in new params field in steps table

Geraint2 committed:
Add support for select list interactive type
Reduce duplicated code for enabling/disabling navigation buttons - more can be done!
When target element is specified in a step but it doesn't exist in the DOM we need to end the tour as we have probably navigated to another page

Geraint2 committed:
Remove redundant use statement

Geraint2 committed:
allow text field to support textarea too

Geraint2 committed:
make the required value field a text area

Geraint2 committed:
remove Factory::getLanguage

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

richard67 commented:
1. The new "params" column has to be added to the extensions.sql for PostgreSQL, too.
2. It needs one update SQL script for each kind of database (MySQl and PotsgreSQL) to add the new column to the table.

richard67 commented:
3. Fix PHP code style errors: We use 4 spaces for each level of indentation in PHP files, not tabs.
4. Fix XML code style.

Geraint2 committed:
PHP code style
Add missing params field into postgres sql file

GeraintEdwards commented:
Thanks @richard67 - I've fixed the missing postgres column, PHP code style and I think I fixed the XML file (the code style guidlines are a bit vague at https://developer.joomla.org/coding-standards/xml.html)

I'm not sure how I would add an update script to add the existing column - I could find no examples in the 5.0dev codebase.  Can you point me in the right direction of suggest an example to look at?

richard67 commented:
> I've fixed the missing postgres column

I can't see that yet.Have you forgotten to push the change?

> I'm not sure how I would add an update script to add the existing column - I could find no examples in the 5.0dev codebase. Can you point me in the right direction of suggest an example to look at?

In the 5.0-dev coce base you won't find much because we removed all the 4.x update SQL scripts. But you can find some examples in the 4.4-dev or 4.3-dev code base:

https://github.com/joomla/joomla-cms/blob/4.4-dev/administrator/components/com_admin/sql/updates/mysql/4.3.2-2023-05-03.sql

https://github.com/joomla/joomla-cms/blob/4.4-dev/administrator/components/com_admin/sql/updates/postgresql/4.3.2-2023-05-20.sql

Geraint2 committed:
PHP code style
Add missing params field into postgres sql file

richard67 commented:
P.S.: The new update SQL script should have a version 5.0.0 and a timestamp newer that the present script "5.0.0-2023-03-11.sql", so I suggest to use "5.0.0-2023-06-21.sql".

richard67 commented on administrator/components/com_guidedtours/src/Model/TourModel.php:
```diff
@@ -355,6 +355,7 @@ public function duplicate(&$pks)
                                 'checked_out_time',
                                 'checked_out',
                                 'language',
+	                            'params',
```
```suggestion
                                'params',
```
There was still one tab among the spaces for indentation. I suggest to use an editor or IDE which shows you tabs and spaces in code.

richard67 commented on administrator/components/com_guidedtours/src/Model/TourModel.php:
```diff
@@ -384,6 +385,7 @@ public function duplicate(&$pks)
                             $db->quoteName('modified'),
                             $db->quoteName('modified_by'),
                             $db->quoteName('language'),
+	                        $db->quoteName('params'),
```
```suggestion
                            $db->quoteName('params'),
```
No tabs.

richard67 commented on administrator/components/com_guidedtours/src/Model/TourModel.php:
```diff
@@ -404,6 +406,7 @@ public function duplicate(&$pks)
                         ParameterType::STRING,
                         ParameterType::INTEGER,
                         ParameterType::STRING,
+	                    ParameterType::STRING,
```
```suggestion
                        ParameterType::STRING,
```
No tabs.

richard67 commented on administrator/components/com_guidedtours/src/Model/TourModel.php:
```diff
@@ -426,6 +429,7 @@ public function duplicate(&$pks)
                                     $date,
                                     $user->id,
                                     $step->language,
+	                                $step->params,
```
```suggestion
                                    $step->params,
```

richard67 commented on administrator/components/com_guidedtours/src/Model/StepsModel.php:
```diff
@@ -231,6 +232,13 @@ public function getItems()
         foreach ($items as $item) {
             $item->title       = Text::_($item->title);
             $item->description = Text::_($item->description);
+
+			if (isset($item->params)) {
+				$params = new Registry($item->params);
+				if (isset($item->params->requiredvalue) && !empty($item->params->requiredvalue)) {
+					$item->params->requiredvalue = Text::_($item->params->requiredvalue);
+				}
+			}
```
```suggestion

            if (isset($item->params)) {
                $params = new Registry($item->params);
                if (isset($item->params->requiredvalue) && !empty($item->params->requiredvalue)) {
                    $item->params->requiredvalue = Text::_($item->params->requiredvalue);
                }
            }
```
No tabs.

Geraint2 committed:
PHP code style
Add missing SQL update scripts

GeraintEdwards commented:
Whoops - postgres change was not pushed  :(

Thanks for the update SQL location.

For some reason my PHPStorm installation hadn't picked up on these tabs when I analysed the code :(

richard67 commented:
> Whoops - postgres change was not pushed :(
> 
> Thanks for the update SQL location.
> 
> For some reason my PHPStorm installation hadn't picked up on these tabs when I analysed the code :(

There are more code style problems with tabs instead of spaces, but I'm busy at work so can't check further.

If you click the "Details" link at the right side of the "continuous-integration/drone/p" step at the bottom of the PR you come to a page with the CI checks. There you select "PHPCS" at the left hand side, then you see the log of the PHP code style check.

Geraint2 committed:
PHP code style
Add missing SQL update scripts

Geraint2 committed:
PHP code style

Geraint2 committed:
PHP code style

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/5.0.0-2023-06-21.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE "#__guidedtour_steps" ADD COLUMN "params" text NOT NULL, /** CAN FAIL **/;
```
```suggestion
ALTER TABLE "#__guidedtour_steps" ADD COLUMN "params" text NOT NULL /** CAN FAIL **/;
```
Comma will cause syntax error.

GeraintEdwards replied:
ARGH!! the perils of cut and paste :(

Geraint2 committed:
bad cut and paste of SQL :(

Quy commented on administrator/components/com_guidedtours/src/Model/StepsModel.php:
```diff
@@ -16,6 +16,7 @@
 use Joomla\Database\DatabaseQuery;
 use Joomla\Database\ParameterType;
 use Joomla\Utilities\ArrayHelper;
+use Joomla\Registry\Registry;
```
```suggestion
use Joomla\Registry\Registry;
use Joomla\Utilities\ArrayHelper;
```

Geraint2 committed:
reoder use statements

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

Geraint2 committed:
Remove unecessary use Rules

joomla-translation-bot committed:
Translation Update (#40731)

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* use 4.3-dev branch for translation bot

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

* Language update

---------

Co-authored-by: Franciska Perisa <9084265+fancyFranci@users.noreply.github.com>
Co-authored-by: Christian Heel <66922325+heelc29@users.noreply.github.com>
Co-authored-by: Stefan Wendhausen <stefan.wendhausen@tec-promotion.de>
Co-authored-by: Olivier Buisard <olivier.buisard@simplifyyourweb.com>
Co-authored-by: Harald Leithner <leithner@itronic.at>

Quy commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -17,6 +17,8 @@ COM_GUIDEDTOURS_FIELD_VALUE_INTERACTIVESTEP_TYPE_BUTTON="Button"
 COM_GUIDEDTOURS_FIELD_VALUE_INTERACTIVESTEP_TYPE_FORM_SUBMIT="Form Submit"
 COM_GUIDEDTOURS_FIELD_VALUE_INTERACTIVESTEP_TYPE_OTHER="Other"
 COM_GUIDEDTOURS_FIELD_VALUE_INTERACTIVESTEP_TYPE_TEXT_FIELD="Text Field"
+COM_GUIDEDTOURS_FIELD_VALUE_INTERACTIVESTEP_TYPE_CHECKBOX_RADIO_FIELD="Checkbox/Radio"
```
Please sort keys in alpha order.

Quy commented on administrator/components/com_guidedtours/forms/step.xml:
```diff
@@ -189,4 +191,42 @@
+			<fieldset name="advanced" label="COM_GUIDEDTOURS_FIELD_ADVANCED_LABEL">
+				<field
+						name="required"
+						type="list"
```
![required](https://github.com/joomla/joomla-cms/assets/368084/5ab0a1ce-8f91-472e-816a-0d9dff74a076)

Please consider making it a switch and changing the label.

GeraintEdwards replied:
No problems about changing to a switcher and changing the fieldset name to something related to required values is a good idea 

I have several other PRs in the pipeline that will add further groups of special/conditional options to this form.

The fieldset(s) don't need to be shown if the fields are not visible because of showon conditions.  The same is true of the containing tab - if there are not visible form fields then the tab should not be shown either.

I'm wondering about putting a PR together to extend the showon script to hide empty fieldsets and tabs.   Without such functionality then the empty tab/fieldset(s) could be confusing to your average user.  What do you think?

Quy replied:
Yes that would be ideal, or move these fields below `Interactive Type` and remove the Options tab.

GeraintEdwards replied:
The problem with putting these fields under the interactive type field is that they are 'params' fields and I don't see an easy way to output these in the form since `echo LayoutHelper::render('joomla.edit.params', $this);`  relies on the params being within fieldsets as far as I can tell.

I'll do a simple showon PR that handles fieldsets and adds a new event - tabs can come later as they are more complicated.

GeraintEdwards replied:
Empty tab and fieldset now respond to showon event triggers.  Ideally there should be an overall showon-processed type event we could respond to but we can wait for that to be available.

heelc29 committed:
[4.3] Update readme (#41088)

sandewt committed:
Color issue (#40435)

Commit:
Change required field to radiobox/selector
Change lanbguage file ordering for label

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

tecpromotion committed:
[4.4] Unification of the spelling in plg_quickicon_eos.sys.ini (#41099)

* Unification of the spelling

* Shortening to Joomla 4

Shortening the spelling from Joomla 4.4 to Joomla 4.

Geraint2 committed:
Language file typo

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

Geraint2 committed:
change fieldset name for required values
Make empty fieldsets and tabs disappear/appear based on showon behaviour of contained fields

Geraint2 committed:
Merge remote-tracking branch 'origin/tour-4' into tour-4

HLeithner commented on build/media_source/com_guidedtours/joomla.asset.json:
```diff
@@ -0,0 +1,34 @@
+  "license": "GPL-2.0-or-later",
+  "assets": [
+    {
+      "name": "com_guidedtours.tour-edit.es5",
```
no es5 assets anymore in j5.0

GeraintEdwards replied:
@HLeithner Sorry - I think I'm being a bit slow here.  This JSON file looks like it follows the same pattern as other core media asset json files.  e.g. com_content

![Screenshot from 2023-07-10 09-45-43](https://github.com/joomla/joomla-cms/assets/839810/23581e10-ac17-424c-95e9-9f5d5acaa22b)

HLeithner replied:
That's legacy code and introducing a new empty file doesn't make sense ;-) so please remove that part of the json. thanks

GeraintEdwards replied:
Removed now

joomla-translation-bot committed:
Translation Update (#41100)

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

Geraint2 committed:
No need to include reference to es5 javascript 5 at all.

Geraint2 committed:
Merge remote-tracking branch 'origin/tour-4' into tour-4

brianteeman committed:
[4.3] typo (#41121)

jurihahn committed:
fix on change JavaScript event on color form field with layout advanced (#41126)

heelc29 committed:
[4.3] check of invalid json 'joomla.asset.json' (#41135)

brianteeman committed:
[4.3] Colour fields doc blocks (#41128)

brianteeman committed:
[4.3] guided tours module count (#40447)

brianteeman committed:
[4.3] Single line comment (#41168)

* [4.3] Single line comment

Codestyle fix

(spotted while doing something else - might as well fix it)

* Update default.php

brianteeman committed:
[4.3] Fix workflow filter (#41167)

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

richard67 committed:
Add missing can fail installer hint (#41184)

Quy commented on administrator/components/com_guidedtours/forms/step.xml:
```diff
@@ -189,4 +191,43 @@
+						filter="safehtml"
+						cols="80"
+						rows="3"
+						default=""
```
```suggestion
					name="required"
					type="radio"
					label="COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_LABEL"
					description="COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_DESC"
					showon=".type:2[AND].interactive_type:2,5,6"
					layout="joomla.form.field.radio.switcher"
					filter="integer"
					validate="options"
					default="1"
					>
					<option value="0">JNO</option>
					<option value="1">JYES</option>
				</field>

				<field
					name="requiredvalue"
					type="textarea"
					label="COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_VALUE_LABEL"
					description="COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_VALUE_DESC"
					showon=".type:2[AND].interactive_type:2,6[AND]required:1"
					filter="safehtml"
					cols="80"
					rows="3"
					default=""
```

Quy commented on administrator/components/com_guidedtours/src/Model/StepsModel.php:
```diff
@@ -233,6 +234,13 @@ public function getItems()
+
+            if (isset($item->params)) {
+                $params = new Registry($item->params);
+                if (isset($item->params->requiredvalue) && !empty($item->params->requiredvalue)) {
```
```suggestion
                if (!empty($item->params->requiredvalue)) {
```

GeraintEdwards committed:
Record currentStepId in sessionStorage when responding to an interactive button - the button may cause a page redirect or form submission (#41155)

MacJoom committed:
Fix for TypeError: count() in webservices when running PHP 8.x  (#40610)

Fedik committed:
[5.0] New onAfterInitialiseDocument event (#40512)

* Document event: BeforeCompileHeadEvent

* Document event: AfterInitialiseDocumentEvent

* Document event, phpcs

* Document event, revert classes

* Document event, cleanup

* Document event, AfterInitialiseDocument Event

Fedik committed:
[5.0] Application event classes (#40522)

* App events: create

* App events: base app event

* App events: add couple event classes

* App events: doc event

* App events: organize

* App events, couple more events

* App events, formatting

* App events, more events

* App events, doc to app

* App events, use same dispatcher

* App events, create to new

---------

Co-authored-by: Harald Leithner <leithner@itronic.at>

HLeithner committed:
[5.0] b/c plugin - Load class aliases earlier and move es5 assets (#41202)

* Move classmap import to the plugin constructor

* Move es5 assets to b/c plugin

---------

Co-authored-by: Brian Teeman <brian@teeman.net>

rajputanuj31 committed:
Added an Api test for contact group field (#41145)

* added a test for contact group field

* updates

---------

rajputanuj31 committed:
Added an api test for content group fields (#41147)

* added an api test for content group fields

* changing file location

---------

richard67 committed:
Fix excluded files from PR #41065 (#41207)

wilsonge committed:
[4.4] Bootstrap 5.3 (#41101)

* Upgrade to bootstrap 5.3

* Restore variable for b/c of com_media overrides

* Fixes for RGB conversions

* Fixes

* Don't use undefined var

* Oops

* Fix path

* New version of system template files

* Move prefix into the variables folder for other scss files

* Remove bs prefix

brianteeman committed:
[5.0] Rename Login with Web Authentication to Login with Passkeys (#41085)

* [5.0] Passkeys

* size

* fgixes

* more

* Revert "Auxiliary commit to revert individual files from c8db802d91b0f009be6f03b3ce24f51e01e2f23d"

This reverts commit 54a7951d408329a3fa24a80145391a94bb25fe29.

* Revert "Auxiliary commit to revert individual files from 8d080250b631945c6284d3b114ea5ec8e31c01b7"

This reverts commit e809ddea25d9ce4f2262fe96dd5faf2cd7410ff7.

* Revert "Auxiliary commit to revert individual files from c8db802d91b0f009be6f03b3ce24f51e01e2f23d"

This reverts commit ee1a4e9b4e936c1bb06fdfc3369875a8b136bb37.

* styleguide

* missed one

* Add image, change login wording following the style guide (#512)

* Add image, change login wording following the style guide

* Add passkeys image for mfa method

---------

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>
Co-authored-by: Harald Leithner <leithner@itronic.at>

alikon committed:
[4.4] update to 5 from cli (#41117)

* Fix Indirect Modification of Overloaded Property

* file_exists

---------

brianteeman committed:
[5.x] SQL field plugin (#41122)

* [5.x] SQL field plugin

* added to the plugin params - n0ot sure why but done for consistency

brianteeman committed:
[5.x] Redirect filter (#41166)

* [5.x] Redirect filter

* camelcase

heelc29 committed:
[5.0] Replace JPATH_PLATFORM with _JEXEC (#41212)

heelc29 committed:
[5.0] Fix cli installation (#41213)

heelc29 committed:
[5.0] replace deprecated access of input property (#41215)

* [5.0] replace deprecated access of input property

* replace Factory::getApplication()

heelc29 committed:
[5.0] uf8 encoding consistent (#41216)

richard67 committed:
[5.0] Update deleted files list in script.php for 5.0.0-alpha3 (#41211)

* Add deleted files and folders from PRs #40664 , #41070 and #41101

* Remove false alarm due to missing upmerge of PR #41101

brianteeman committed:
[[5.x]Compatibility Plugin (BC) added to system info (#41209)

* [5.x]Compatibility Plugin (BC) added to system info

It's going to be more and more important to be able to know if the Backward Compatibility Plugin is enabled or not.

This PR adds it to the list of information displayed in the system information page

* long name

* Add parameters to Combat sysinfo

* cs

* Update SysinfoModel.php

---------

Co-authored-by: Harald Leithner <leithner@itronic.at>

brianteeman committed:
[5.x] System - Backward Compatibility params (#41214)

* [5.x] System - Backward Compatibility params

Display the params correctly

* update the language strings

* description

heelc29 committed:
[5.0] replace deprecated access of input property (#41219)

Fedik committed:
Fix error rendering with unclosed Output buffers (#41188)

shazmasiddiqui committed:
[5.0] [GSoC22] Rich Snippets - schema.org implementation (#41151)

* Update drone signature

* Base Implementation of schema.org

* Adds language strings

* Fixed absolute path for schemaorg.xml

* Bloposting plugin

* Book plugin

* Event Plugin

* Updates language files

* Adds organization schema

* Fix file line end

* Add empty line at end of file

* Add empty line at end of file

* Update provider.php

Added empty lines

* Update Book.php

Added empty lines

* Added empty line at the end of file

* Resolving conflicts

* Update 4.2.3-2022-12-10.sql

* Update 4.2.3-2022-12-10.sql

* Update base.sql

* Update base.sql

* Update 4.2.3-2022-12-10.sql

* Update 4.2.3-2022-12-10.sql

* Update provider.php

* Update Event.php

* Adds Person plugin

* Adds empty line

* Adds empty line

* Adds empty line at the end of file

* Update provider.php

* Moved xml outside of src folder and load via trait

* Remove $supportFunctionality

* Move onSchemaPrepareData to trait

* Move onSchemaPrepareData to trait

* Moved getSubscribedEvents to trait

* Move onSchemaAfterSave to trait

* Adds recipe plugin

* Use EventInterface as type hint

* Move XML outside of src folder for schema.org system plugin

* Updated file structure

* Update plg_schemaorg_recipe.ini

* Update duration.xml

* Update provider.php

* cleanups

* Adds schemaorg in contacts component

* Add empty line

* SQL fixes

- Consistent SQL for new installations and updates
- Larger data type for id and itemId columns
- No display width for integer columns on MySQL
- Use order number for schemaorg plugins
- Fix order numbers of system plugins on new installation
- Fix character set and collation for MySQL on new installation
- Align code style to existing SQL

* One more CS fix

* Alphabetically sort new schemaorg plugins

* Update ContactComponent.php

* Fix language files

* version updated

* Deleting rich snippets from components/com_content folder

* Removed rich snippets from contact component

* Removes rich snippets from layout folder

* minor fixes

* Update drone hash

* Fixed phpcs

* Fix code style

* Improve information for schema and require base configuration

* Fix basetype parameter

* Added base type in output

* Remove debug output

* Add breadcrumb and finder implementation

* Add image and social media

* Check for context if schema is supported for the extension

* CS Fix

* CS fixes

* CS fix

* Fix deploy version

* Improve data loading

* Improve trait

* Fix broken variables

* CS fix

* CS fix

* CS fix

* CS fix

* Fix from type

* Cleanup plugins

* Add base structure of blogposting

* Add JobPosting installer

* Fix CS

* Fix new line

* Form fields for jobposting

* Fix structure

* Provide form for schema.org JoPosting

* Fix linebreaks

* Remove unused namespace

* Changed schema structure and fixed the customCleanup

* Added generic field

* Changed input type for Nutrition fields to text

* Update .drone.yml

* Update plugins/schemaorg/jobposting/forms/schemaorg.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/jobposting/forms/schemaorg.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/jobposting/jobposting.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/jobposting/jobposting.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/src/Extension/Recipe.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/src/Extension/Recipe.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/src/Extension/Person.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/src/Extension/Recipe.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/book/src/Extension/Book.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/src/Extension/Person.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/src/Extension/Person.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/jobposting/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/jobposting/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/language/en-GB/plg_schemaorg_event.sys.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/src/Extension/Person.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/components/com_contact/src/Extension/ContactComponent.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/src/Extension/Person.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/components/com_content/src/Extension/ContentComponent.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/language/en-GB/plg_schemaorg_blogposting.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/src/Extension/Recipe.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/language/en-GB/plg_schemaorg_blogposting.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/organization.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/src/Extension/Organization.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/src/Extension/Organization.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/src/Extension/Organization.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/person.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/organization/src/Extension/Organization.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/src/Extension/Schemaorg.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/person/person.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/recipe.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/recipe.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/schemaorg/recipe/src/Extension/Recipe.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/schemaorg.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update plugins/system/schemaorg/schemaorg.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update libraries/src/Form/Field/SchemaorgComponentSectionsField.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update libraries/src/Form/Field/SchemaorgComponentSectionsField.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update libraries/src/Form/Field/SchemaorgComponentSectionsField.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update libraries/src/Schemaorg/SchemaorgPluginTrait.php

Co-authored-by: Brian Teeman <brian@teeman.net>

* Update administrator/language/en-GB/plg_schemaorg_blogposting.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

* Fix base.sql duplicate system plugins

* Add back schemaorg plugins

* Cleanup and com_content schema fallback integration

* Add contact fallback schema

* Add contact featured bc schema

* Apply suggestions from @brianteeman

Co-authored-by: Brian Teeman <brian@teeman.net>

* Rename update files

* Fix wrong DB call

* Add check for existing schemas

* Apply suggestions from code review

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>
Co-authored-by: Brian Teeman <brian@teeman.net>

* CS

* Apply suggestions from code review

Co-authored-by: Brian Teeman <brian@teeman.net>

* Apply suggestions from code review

Co-authored-by: Brian Teeman <brian@teeman.net>

---------

Co-authored-by: Benjamin Trenkle <bembelimen@users.noreply.github.com>
Co-authored-by: Olivier Buisard <olivier.buisard@simplifyyourweb.com>
Co-authored-by: anuragteapot <anuragvns1999@gmail.com>
Co-authored-by: Benjamin Trenkle <benjamin.trenkle@wicked-software.de>
Co-authored-by: Richard Fath <richard67@users.noreply.github.com>
Co-authored-by: Timo Feuerstein <timo.feuerstein@renolit.com>
Co-authored-by: Brian Teeman <brian@teeman.net>
Co-authored-by: Richard Fath <richard.fath@t-online.de>
Co-authored-by: Harald Leithner <leithner@itronic.at>

brianteeman committed:
[5.x] schemaorg plugins mark as core (#41227)

This PR adds the missing code so that the new system and schemaorg plugins are listed as core extensions in the extension manager

HLeithner committed:
Revert part of #40429

There was an error loading the language base path using the framework OutputFilter.

HLeithner committed:
Joomla! 5.0.0 Alpha 3

HLeithner committed:
Reset to dev

brianteeman committed:
[5.0] schemaorg missing namespace (#41237)

Removes the code that disables phpcs checking for a missing namespace as it is present.

fixes some codestyle at the same time

codereview and if it passes phpcs checks

brianteeman committed:
[5.0] schemaorg missing copyright (#41234)

* [5.0] schemaorg missing copyright

Adds missing copyright statement

code review

First reported
https://github.com/joomla/joomla-cms/pull/41151#discussion_r1261749182
https://github.com/joomla/joomla-cms/pull/41151#discussion_r1261749351

* Update libraries/src/Schemaorg/SchemaorgServiceTrait.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

---------

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

brianteeman committed:
[5.0] Schemaorg missing string (#41232)

Every .sys.ini file has an XML_DESCRIPTION string and every .ini file has the same string. This must have been done for a reason and this one language file is missing the string

Core Review

(originally reported and dismissed https://github.com/joomla/joomla-cms/pull/41151#discussion_r1261731953)

brianteeman committed:
[5.0] schemaorg xml (#41235)

This is just codestyle changes removing extra spaces and new lines which came from my misformatted code suggestions

And a missing xml definition in a file header

code review

brianteeman committed:
[5.0] schemaorg spelling alrady\already (#41241)

code review only

heelc29 committed:
[5.0] update language strings of eos quickicon plugin (#41248)

brianteeman committed:
[5.0] scheaeorg return array (#41240)

phan reports Type Mismatch - hope I fixed it correctly

plugins/schemaorg/blogposting/src/Extension/Blogposting.php:55 PhanTypeMismatchReturn Returning $this->cleanupDate($schema, ['datePublished','dateModified']) of type bool but customCleanup() is declared to return array
plugins/schemaorg/book/src/Extension/Book.php:55 PhanTypeMismatchReturn Returning $this->cleanupDate($schema, ['datePublished']) of type bool but customCleanup() is declared to return array
plugins/schemaorg/event/src/Extension/Event.php:55 PhanTypeMismatchReturn Returning $this->cleanupDate($schema, ['startDate']) of type bool but customCleanup() is declared to return array
plugins/schemaorg/jobposting/src/Extension/JobPosting.php:57 PhanTypeMismatchReturn Returning $schema of type bool but customCleanup() is declared to return array
plugins/schemaorg/recipe/src/Extension/Recipe.php:61 PhanTypeMismatchReturn Returning $schema of type bool but customCleanup() is declared to return array

heelc29 committed:
[5.0] schemaorg utf8 encoding consistent (#41247)

heelc29 committed:
[5.0] schemaorg use injected application from plugin provider (#41246)

heelc29 committed:
[5.0] schemaorg fix deploy version (#41245)

heelc29 committed:
[5.0] schemaorg replace JPATH_PLATFORM with _JEXEC (#41244)

* [5.0] schemaorg replace JPATH_PLATFORM with _JEXEC

* add copyright

heelc29 committed:
[5.0] schemaorg language files (#41243)

heelc29 committed:
[5.0] schemaorg sort imports (#41242)

HLeithner committed:
Fix missing legacy schema for category,featured,archive view (#41250)

Unescape JSON output

richard67 committed:
Fix XML syntax error from PR #41247 (#41257)

sergeytolkachyov committed:
[5.0] Update bootstrap modal layout - add modalCss optional parameter (#40650)

* Update layouts bootstrap modal main.php

Add an optional parameter 'modalCss' to $params array

* Fixed code style

* Update code style

* Update layouts/libraries/html/bootstrap/modal/main.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

---------

Co-authored-by: Quy <quy@nomonkeybiz.com>
Co-authored-by: Robert Deutz <rdeutz@googlemail.com>

brianteeman committed:
[4.3] Guided Tours disabled plugin (#40449)

heelc29 committed:
[5.0] LICENSE.txt (#41277)

SniperSister committed:
Fix com_config dispatcher (#41262)

iamrobert committed:
Validation Issue (#41285)

![image](https://i.imgur.com/uJ4o8tc.png)

Block elements are not allowed inside inline elements. 

The button tag is an inline element, so the inside div isn't allowed. 

Switching to span seems to resolve the issue.

brianteeman committed:
[4.3] Display Contact quickicon (#41282)

Fixes a typo introduced with #40715 that prevented the contacts quickicon being displayed

Co-authored-by: Quy <quy@nomonkeybiz.com>

wilsonge committed:
[4.4] Fix minor UI issues caused by bootstrap 5.3 version bump (#41286)

* Fix missing variable maps in atum

* Fix color on login page

* Fix codestyle

HLeithner committed:
[5.0] Move the compatibility plugin from system to behaviour folder (#41276)

* Move the compatibility plugin from system to behaviour folder

This allows system plugins using the JPlugin class to still work with joomla 5.0 compatibility enabled

* Fix Language string

* Fix SQL update scripts for compat plugin

brianteeman committed:
[4.3] com_messages typo (#41295)

HLeithner committed:
Remove class aliases to non existing classes (#41302)

JClientLdap and JLDAP, dependency has been removed

heelc29 committed:
[5.0] fix compat plugin folder change (#41310)

wilsonge committed:
Fix when deprecation was introduced

See https://developer.joomla.org/joomlacode-archive/issue-31727.html

brianteeman committed:
[4.4] Deprecation string (#41210)

https://github.com/joomla/joomla-cms/pull/38657

bembelimen committed:
[5.0] schema.org cleanup (#41260)

* Add json cleanup

* CS fix

* Fix deprecation and file name

* Fix deprecations

* Apply suggestions from code review

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>
Co-authored-by: Brian Teeman <brian@teeman.net>

* Update libraries/src/Schemaorg/SchemaorgPrepareDurationTrait.php

* Update libraries/src/Schemaorg/SchemaorgPrepareDurationTrait.php

* Apply suggestions from code review

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

* Fix namespace

* Activate cache by default

---------

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>
Co-authored-by: Brian Teeman <brian@teeman.net>

heelc29 committed:
[4.3] PHP 8.2 Dynamic property - com_config request controller (#41292)

brianteeman committed:
[4.3] Content Archive a11y (#41329)

Hackwar committed:
[4.3] Finder: Fixing tokenisation for chinese text (#41275)

heelc29 committed:
[5.0] schemaorg fix user parameter (#41269)

heelc29 committed:
[5.0] schemaorg fix schema forms (#41323)

fix option value
add missing language keys

brianteeman committed:
[5.0] Archive content schemaorg (#41344)

brianteeman committed:
[5.0] link docblock (#41178)

* input filter clean

* second pass

* third pass

* final pass

* cleanup

* indent

heelc29 committed:
[5.0] build tools creates wrong folder (#41321)

* [5.0] build tools creates wrong folder

* update/sync incompatible.html

---------

Co-authored-by: Quy <quy@nomonkeybiz.com>

alikon committed:
[5][com_joomlaupdate] Deprecated: Creation of dynamic property php 8.2 (#41118)

* Deprecated: Creation of dynamic property 8.2

* Deprecated: Creation of dynamic property

* update:extensions:check

* -redeclare

* noclueotherthan

---------

Co-authored-by: Harald Leithner <leithner@itronic.at>

heelc29 committed:
[5.0] schemaorg format json-ld (#41271)

* [5.0] schemaorg format json-ld

do not escape '/' (breadcrumbs)

* [5.0] schemaorg format json-ld

pretty print when debug enabled

* update

---------

Co-authored-by: Benjamin Trenkle <bembelimen@users.noreply.github.com>
Co-authored-by: Quy <quy@nomonkeybiz.com>

heelc29 committed:
[5.0] fix joomla no download url (#41364)

Fedik committed:
[5.0] Update JSON flags for rendering slashes and unicode, and debugging (#41330)

* Update JSON flags for rendering slashes and unicode, and debugging

* revert schema

* shorten

Fedik committed:
[5.0] TinyMCE highlighter plugin with CodeMirror 6 (#41289)

* joomla-highlighter

* joomla-highlighter

* fixes

* joomla-highlighter

* clean

* clean

* focus

* Fix styling

* Tab trap

Geraint2 committed:
Merge remote-tracking branch 'upstream/5.0-dev' into tour-4

# Conflicts:
#	build/media_source/plg_system_guidedtours/js/guidedtours.es6.js

Geraint2 committed:
Merge head correctly

GeraintEdwards committed:
Update administrator/components/com_guidedtours/forms/step.xml

Co-authored-by: Quy <quy@nomonkeybiz.com>

GeraintEdwards committed:
Update administrator/components/com_guidedtours/src/Model/StepsModel.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

Geraint2 committed:
Merge remote-tracking branch 'upstream/5.0-dev' into tour-4

# Conflicts:
#	administrator/language/en-GB/com_guidedtours.ini

Geraint2 committed:
Merge and resolve conclict with head

Geraint2 committed:
Merge remote-tracking branch 'origin/tour-4' into tour-4

GeraintEdwards committed:
Merge branch '5.0-dev' into tour-4

HLeithner commented:
This pull request has been automatically rebased to 5.1-dev.

obuisard committed:
Merge branch '5.1-dev' into tour-4

obuisard committed:
Update StepTable.php

Reorder of use statements

sdwjoomla commented:
I have tested this item :white_check_mark: successfully on 20c8ac4e29a5f5ab2bce4d72a2db4eddfb20fafc<br /><br />Worked as expected for the toggle, dropdown and required text.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40994">issues.joomla.org/tracker/joomla-cms/40994</a>.</sub>

richard67 commented on administrator/components/com_admin/sql/updates/mysql/5.0.0-2023-06-21.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE `#__guidedtour_steps` ADD COLUMN `params` text NOT NULL /** CAN FAIL **/;
```
The update SQL scripts need to be renamed to 5.1.0 with a date newer than 2024-01-04, e.g. to „5.1.0-2024-02-01.sql“.

garstud commented:
I have tested this item ✅ successfully with the Prebuilt package Joomla_5.1.1-alpha1-dev+pr.40994-Development-Full_Package (under Chrome and Firefox)

obuisard committed:
Update step.xml

The textarea field cannot have options

obuisard committed:
Merge branch '5.1-dev' into tour-4

obuisard commented:
This PR is being discussed during our Guided Tours meetings and suggested improvements are under way for:
- the wording of the parameters/naming of the fields
- adding a message (note field) to the Options tab to force the tab to exist always (when the options tab is open and a new step is created, the user is greeted with a blank section because the tab is missing).

An issue was found when a radio button's value is NO and is required to be set to YES to go further:
when the radio button is set to YES (the 'next' button becomes available) then set to NO again, the 'next' button is not going back to being disabled, the user can go through to the next step even though the radio button is set to NO.

That said, it would be nice to force a radio button to have a NO value before being able to get to the next step (we should allow the required value field to be visible for radio/checkbox and specify 0 or 1 as the required value).

obuisard committed:
Update edit.php

space correction

obuisard committed:
Update com_guidedtours.ini

obuisard committed:
Update com_guidedtours.ini

Corrected order

obuisard committed:
Update step.xml

Added a note field to allow the options tab to be present, always and to give context to the user.
Changed language key names and fieldset name (trying to keep consistency with other parts in Joomla)

obuisard committed:
Update step.xml

Added showon to note field

obuisard committed:
Update step.xml

Simplified showon in note field

obuisard committed:
Merge branch '5.1-dev' into tour-4

garstud commented:
I have tested this item :white_check_mark: successfully on 20c8ac4e29a5f5ab2bce4d72a2db4eddfb20fafc<br /><br />Ok for interactives Steps (RadioButton, text, list) and Ok for the Options Tab with info text.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40994">issues.joomla.org/tracker/joomla-cms/40994</a>.</sub>

sdwjoomla commented:
I have tested this item :white_check_mark: successfully on 20c8ac4e29a5f5ab2bce4d72a2db4eddfb20fafc<br /><br />Retested successfully after changes<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40994">issues.joomla.org/tracker/joomla-cms/40994</a>.</sub>

AboutTimeIT commented:
I have tested this item :white_check_mark: successfully on 20c8ac4e29a5f5ab2bce4d72a2db4eddfb20fafc<br /><br />Test was successfull.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40994">issues.joomla.org/tracker/joomla-cms/40994</a>.</sub>

Quy commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -74,6 +78,10 @@ COM_GUIDEDTOURS_TOURS_LIST="Guided Tours"
 COM_GUIDEDTOURS_TOURS_LIST_EMPTYSTATE_CONTENT="Create a tour to make it functional."
 COM_GUIDEDTOURS_TOURS_LIST_EMPTYSTATE_TITLE="No tours have been created yet."
+COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_DESC="Enable if the user is required to provide a value, activate a radio button or check a box to move forward to the next step of the tour"
+COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_LABEL="Value Is Required"
```
```suggestion
COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_LABEL="Required"
```

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Quy <quy@nomonkeybiz.com>

richard67 commented:
@obuisard Please rename the update SQL scripts like I suggested in my review comment.

obuisard committed:
Rename 5.0.0-2023-06-21.sql to 5.1.0-2024-02-10.sql

obuisard committed:
Rename 5.0.0-2023-06-21.sql to 5.1.0-2024-02-10.sql

obuisard commented:
> @obuisard Please rename the update SQL scripts like I suggested in my review comment.

Thanks for the reminder Richard @richard67, I totally missed that one :-)

richard67 commented:
> Thanks for the reminder Richard @richard67, I totally missed that one :-)

@obuisard Maybe my mistake because I just saw I had not finished that review with that comment so the review was still ongoing and maybe only me could see it.

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40994">issues.joomla.org/tracker/joomla-cms/40994</a>.</sub>

richard67 committed:
Merge branch '5.1-dev' into tour-4

obuisard commented:
> > Thanks for the reminder Richard @richard67, I totally missed that one :-)
> 
> @obuisard Maybe my mistake because I just saw I had not finished that review with that comment so the review was still ongoing and maybe only me could see it.

No, no, my fault, I should know better :-)

LadySolveig committed:
Merge branch '5.1-dev' into tour-4

richard67 commented on administrator/components/com_admin/sql/updates/mysql/5.1.0-2024-02-10.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE `#__guidedtour_steps` ADD COLUMN `params` text NOT NULL /** CAN FAIL **/;
```
```suggestion
ALTER TABLE `#__guidedtour_steps` ADD COLUMN `params` text NULL /** CAN FAIL **/;
```
We shouldn't use the NOT NULL constraint for the new columns as that can make problems for updates, or at least can make it more complicated on PostgreSQL. The `#__categories` table also allows null values for its `params` column, so we would be consistent with that.

obuisard replied:
Thanks Richard

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/5.1.0-2024-02-10.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE "#__guidedtour_steps" ADD COLUMN "params" text NOT NULL /** CAN FAIL **/;
```
```suggestion
ALTER TABLE "#__guidedtour_steps" ADD COLUMN "params" text NULL /** CAN FAIL **/;
```
See previous review comment.

richard67 commented on installation/sql/mysql/extensions.sql:
```diff
@@ -1022,6 +1022,7 @@ CREATE TABLE IF NOT EXISTS `#__guidedtour_steps` (
   `checked_out` int unsigned,
   `language` varchar(7) NOT NULL,
   `note` varchar(255) NOT NULL DEFAULT '',
+  `params` text NOT NULL,
```
```suggestion
  `params` text,
```
See previous review comments.

richard67 commented on installation/sql/postgresql/extensions.sql:
```diff
@@ -989,6 +989,7 @@ CREATE TABLE IF NOT EXISTS "#__guidedtour_steps" (
   "checked_out" integer,
   "language" varchar(7) DEFAULT '' NOT NULL,
   "note" varchar(255) DEFAULT '' NOT NULL,
+  "params" text NOT NULL,
```
```suggestion
  "params" text,
```
See previous review comments.

richard67 commented:
@GeraintEdwards @obuisard We should not use s NOT NULL constraint for the new column, even if other tables have that for their `params` column. But e.g. the `#__categories` table allows NULL values for its `params` column.

I have made corresponding review comments with change suggestions for that.

The reason is following:

When we add the new column to the table on updates with a NOT NULL constraint, it works on MySQL and MariaDB because that implicitly sets the value to an empty string for that new column.

But on PostgreSQL that does not work. We have to add the column first without the NOT NULL constraint, then update the existing records, and then add the NOT NULL CONSTRAINT. If this fails for some reason, the database fixer will not be able to fix that because it will only run the ALTER TABLE but not the UPDATE statement. That may cause issues.

Therefore it's the best to allow NULLs, and as far as I can see in the PHP code of this PR it handles that already. But this has to be tested in order to be safe.

obuisard committed:
Update administrator/components/com_admin/sql/updates/mysql/5.1.0-2024-02-10.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

obuisard committed:
Update installation/sql/mysql/extensions.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

obuisard committed:
Update administrator/components/com_admin/sql/updates/postgresql/5.1.0-2024-02-10.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

obuisard committed:
Update installation/sql/postgresql/extensions.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

obuisard commented:
> @GeraintEdwards @obuisard We should not use a NOT NULL constraint for the new column, even if other tables have that for their `params` column. But e.g. the `#__categories` table allows NULL values for its `params` column.
> 
> I have made corresponding review comments with change suggestions for that.
> 
> The reason is following:
> 
> When we add the new column to the table on updates with a NOT NULL constraint, it works on MySQL and MariaDB because that implicitly sets the value to an empty string for that new column.
> 
> But on PostgreSQL that does not work. We have to add the column first without the NOT NULL constraint, then update the existing records, and then add the NOT NULL CONSTRAINT. If this fails for some reason, the database fixer will not be able to fix that because it will only run the 2 ALTER TABLE statements but not the UPDATE statement between them, so the 2nd ALTER TABLE will always fail.
> 
> Therefore it's the best to allow NULLs, and as far as I can see in the PHP code of this PR it handles that already. But this has to be tested in order to be safe.

Thank you so much Richard @richard67 for the explanation, it does make perfect sense. Really appreciate your insight on this!

Quy commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -74,6 +78,10 @@ COM_GUIDEDTOURS_TOURS_LIST="Guided Tours"
 COM_GUIDEDTOURS_TOURS_LIST_EMPTYSTATE_BUTTON_ADD="Add your first tour"
 COM_GUIDEDTOURS_TOURS_LIST_EMPTYSTATE_CONTENT="Create a tour to make it functional."
 COM_GUIDEDTOURS_TOURS_LIST_EMPTYSTATE_TITLE="No tours have been created yet."
+COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_DESC="Enable if the user is required to provide a value, activate a radio button or check a box to move forward to the next step of the tour"
+COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_LABEL="Required"
+COM_GUIDEDTOURS_TYPE_INPUT_REQUIREDVALUE_DESC="The exact value to be entered, including case and punctuation, to move forward to the next step (if the target is a list of items, use the value of the select's option element)"
```
```suggestion
COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_DESC="Enable if the user is required to provide a value, activate a radio button or check a box to move forward to the next step of the tour."
COM_GUIDEDTOURS_TYPE_INPUT_REQUIRED_LABEL="Required"
COM_GUIDEDTOURS_TYPE_INPUT_REQUIREDVALUE_DESC="The exact value to be entered, including case and punctuation, to move forward to the next step (if the target is a list of items, use the value of the select's option element)."
```

richard67 replied:
> use the value of the select's option element

Shouldn’t that better be „use the value of the selected option's element“?

obuisard replied:
> > use the value of the select's option element
> 
> Shouldn’t that better be „use the value of the selected option's element“?

Here, we are talking about the select element, as HTML select

Quy commented on build/media_source/plg_system_guidedtours/js/guidedtours.es6.js:
```diff
@@ -127,19 +147,66 @@ function addStepToTourButton(tour, stepObj, buttons) {
-              primaryButton.classList.add('disabled');
-            }
+          // Check to see if the 'next' button should be enabled before showing the step based on being required or
+          // matcching the required value
```
```suggestion
          // matching the required value
```

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Quy <quy@nomonkeybiz.com>

obuisard committed:
Update build/media_source/plg_system_guidedtours/js/guidedtours.es6.js

Co-authored-by: Quy <quy@nomonkeybiz.com>

Quy commented:
![40994-type](https://github.com/joomla/joomla-cms/assets/368084/497a20c2-bdc1-42a8-b444-c80d0dd6bdd8)
Expand the description to include `Checkbox/Radio` and `Select List`?

obuisard committed:
Update com_guidedtours.ini

Improved description of the interactive type field

obuisard commented:
> ![40994-type](https://private-user-images.githubusercontent.com/368084/305757557-497a20c2-bdc1-42a8-b444-c80d0dd6bdd8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzMDUxMDcsIm5iZiI6MTcwODMwNDgwNywicGF0aCI6Ii8zNjgwODQvMzA1NzU3NTU3LTQ5N2EyMGMyLWJkYzEtNDJhOC1iNDQ0LWM4MGQwZGQ2YmRkOC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxOVQwMTA2NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNDQwZTMyYjBiM2U0NTVhN2Y3OGNhNWQzNjA0ZDlmNTFmMjFkMTc1ODA2OGQyMmU5YzE2YTM3YjNhMjViMTE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ikAGgTiu5f-j1HNI1mMeI-8KGBCPOZ2ybXzK7q95oiw) Expand the description to include `Checkbox/Radio` and `Select List`?

I made a note of that, thanks! The whole description needs to be updated.

> ![40994-type](https://private-user-images.githubusercontent.com/368084/305757557-497a20c2-bdc1-42a8-b444-c80d0dd6bdd8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzMDU2ODksIm5iZiI6MTcwODMwNTM4OSwicGF0aCI6Ii8zNjgwODQvMzA1NzU3NTU3LTQ5N2EyMGMyLWJkYzEtNDJhOC1iNDQ0LWM4MGQwZGQ2YmRkOC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxOVQwMTE2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03OTIwY2FjZGNiOTdiZjc4NTJjNWI5MzYyMzU4YzAwMDNkMWQ2NTZjNjc3MzQyZTBhNmYxNzJlZDc0NTMzNTUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QC9QJMaJp3GeIBcM-yCgKl6HxykFchPv01YnC2qVKZA) Expand the description to include `Checkbox/Radio` and `Select List`?

Good point, thanks!

Kostelano commented:
@GeraintEdwards 

Good afternoon. I’m not sure yet, I’m still figuring it out, but...

When creating a step, I select the interactive and checkbox/radio, but I don't see the target value in the Target Value Options tab (similar to other interactives).

![Screenshot_1](https://github.com/joomla/joomla-cms/assets/8440661/b4eb4726-41c1-4fed-b78f-110d8df3beb6)

![Screenshot_2](https://github.com/joomla/joomla-cms/assets/8440661/dcfd7f41-eaa8-48cc-94e9-d3128e147a13)

Is this expected behavior or a bug?

GeraintEdwards commented:
> @GeraintEdwards
> 
> Good afternoon. I’m not sure yet, I’m still figuring it out, but...
> 
> When creating a step, I select the interactive and checkbox/radio, but I don't see the target value in the Target Value Options tab (similar to other interactives).
> 
> Is this expected behavior or a bug?
> 
> As far as I understand, from the selection list I can NOT select everything, but only some items.

At present the code allows one selector - the specific checkbox/radiobox and the required option needs this to be checked.  I guess it may make sense to generalise this so allow for the requirement that the checkbox/radiobox to be unchecked and also to allow require multiple options to be checked (or selected in select list fields).  This is more complex to implement as the values are not available to us when editing the steps.

It would, of course, be really neat if we could create/edit the tour and its steps in-situ so that we could pick up the values from the form in real time using javascript.  But I think that is a step to advanced for now.
